examples/mem-events.rs

```diff
@@ -0,0 +1,104 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Instant;
+
+use clap::{App, Arg, ArgMatches};
+use colored::*;
+use env_logger;
+use kvmi::{KVMiPageAccess};
+
+use microvmi::api::*;
+
+const PAGE_SIZE: usize = 4096;
+
+fn parse_args() -> ArgMatches<'static> {
+    App::new(file!())
+        .version("0.3")
+        .author("Mathieu Tarral")
+        .about("Watches memory VMI events")
+        .arg(Arg::with_name("vm_name").index(1).required(true))
+        .get_matches()
+}
+
+fn get_pagefaults(drv: &mut Box<dyn Introspectable>, enabled: bool) {
+    drv.pause().expect("Failed to pause VM");
+
+    let intercept = InterceptType::Pagefault;
+    let status_str = if enabled { "Enabling" } else { "Disabling" };
+    println!("{} memory events", status_str);
+    for vcpu in 0..drv.get_vcpu_count().unwrap() {
+        drv.toggle_intercept(vcpu, intercept, enabled)
+            .expect(&format!("Failed to enable page faults"));
+    }
+
+    drv.resume().expect("Failed to resume VM");
+}
+
+fn main() {
+    env_logger::init();
+
+    let matches = parse_args();
+
+    let domain_name = matches.value_of("vm_name").unwrap();
+
+    // set CTRL-C handler
+    let running = Arc::new(AtomicBool::new(true));
+    let r = running.clone();
+    ctrlc::set_handler(move || {
+        r.store(false, Ordering::SeqCst);
+    })
+    .expect("Error setting Ctrl-C handler");
+
+    println!("Initialize Libmicrovmi");
+    let mut drv: Box<dyn Introspectable> = microvmi::init(domain_name, None);
+    println!("Listen for memory events...");
+    // record elapsed time
+    let start = Instant::now();
+    get_pagefaults(&mut drv, true);
+    let mut i: u64 = 0;
+
+    //Code snippet to get page fault
```

Comment:
@arnabcs17b006 to trigger some pagefaults in KVM, you need
1. enable PF events
2. remove a permission access (r/w/x) on a page which needs it

Here I can see that you do 1. , but then you do:

- curr_addr = 0
- var = PageAccessR
- buffer = [0; 4096]
- get_page_access(0)
- set_page_access(0, new accesses)
- read_physical(0)

There is nothing happened at address 0.
it's a page that's marked not readable, not executable, and not writable by the OS, in case of a null pointer dereference (I believe).

the problem of libkvmi is that i needs physical addresses to work with.
this example in Libvmi uses virtual address translation to find back the physical address of the current RIP, get the frame, and remove the execute permissions, triggering some pagefaults:
https://github.com/KVM-VMI/libvmi/blob/kvmi/examples/mem-event-example.c

Maybe you could iterate get the max gfn, and iterate over every frame to remove the execute permissions, over the entire RAM ?
What do you think @rageagainsthepc  ?

Reply:
Actually we can mimik how `kvmi-test.c` does:
https://github.com/KVM-VMI/kvm/blob/kvmi/tools/kvm/kvmi/kvmi-test.c#L207

They change the permission of the current process's page tables, to get some events